File: src/main/java/io/github/intellij/dlanguage/stubs/types/StaticConstructorStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import io.github.intellij.dlanguage.stubs.DlangStaticConstructorStub;
import io.github.intellij.dlanguage.psi.DLanguageStaticConstructor;
import io.github.intellij.dlanguage.psi.impl.DLanguageStaticConstructorImpl;
import io.github.intellij.dlanguage.stubs.DlangStaticConstructorStub;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class StaticConstructorStubElementType extends DStubElementType<DlangStaticConstructorStub, DLanguageStaticConstructor> {
    public StaticConstructorStubElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DLanguageStaticConstructor createPsi(@NotNull final DlangStaticConstructorStub stub) {
        return new DLanguageStaticConstructorImpl(stub, this);
    }

    @NotNull
    @Override
    public DlangStaticConstructorStub createStub(@NotNull final DLanguageStaticConstructor psi, final StubElement parentStub) {
        return new DlangStaticConstructorStub(parentStub, this);
    }

    @NotNull
    @Override
    public DlangStaticConstructorStub deserialize(@NotNull final StubInputStream dataStream, final StubElement parentStub) throws IOException {
        return new DlangStaticConstructorStub(parentStub, this);
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/StaticConstructorStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import io.github.intellij.dlanguage.psi.DLanguageStaticConstructor
import io.github.intellij.dlanguage.psi.impl.DLanguageStaticConstructorImpl
import io.github.intellij.dlanguage.stubs.DLanguageStaticConstructorStub
import java.io.IOException

class StaticConstructorStubElementType(debugName: String) :
    DStubElementType<DLanguageStaticConstructorStub, DLanguageStaticConstructor>(debugName) {
    override fun createPsi(stub: DLanguageStaticConstructorStub): DLanguageStaticConstructor {
        return DLanguageStaticConstructorImpl(stub, this)
    }

    override fun createStub(psi: DLanguageStaticConstructor, parentStub: StubElement<*>?): DLanguageStaticConstructorStub {
        return DLanguageStaticConstructorStub(parentStub, this)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>?): DLanguageStaticConstructorStub {
        return DLanguageStaticConstructorStub(parentStub, this)
    }
}
